plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveNodePartitioningProvider.java

```diff
@@ -42,15 +59,30 @@ public BucketFunction getBucketFunction(
             int bucketCount)
     {
         HivePartitioningHandle handle = (HivePartitioningHandle) partitioningHandle;
-        List<HiveType> hiveTypes = handle.getHiveTypes();
-        return new HiveBucketFunction(handle.getBucketingVersion(), bucketCount, hiveTypes);
+        List<HiveType> hiveBucketTypes = handle.getHiveBucketTypes();
+        if (handle.getHivePartitionTypes().isEmpty()) {
+            return new HiveBucketFunction(handle.getBucketingVersion(), bucketCount, hiveBucketTypes);
+        }
+        checkArgument(
+                partitionChannelTypes.size() == hiveBucketTypes.size() + handle.getHivePartitionTypes().get().size(),
+                "partitionChannelTypes must include hiveBucketTypes and hivePartitionTypes");
+        return new HivePartitionedBucketFunction(
+                handle.getBucketingVersion(),
+                handle.getBucketCount(),
+                hiveBucketTypes,
+                partitionChannelTypes.subList(hiveBucketTypes.size(), partitionChannelTypes.size()),
+                typeOperators,
+                bucketCount);
     }
 
     @Override
     public ConnectorBucketNodeMap getBucketNodeMap(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
     {
         HivePartitioningHandle handle = (HivePartitioningHandle) partitioningHandle;
-        return createBucketNodeMap(handle.getBucketCount());
+        if (handle.getHivePartitionTypes().isEmpty()) {
+            return createBucketNodeMap(handle.getBucketCount());
+        }
+        return createBucketNodeMap(nodeManager.getRequiredWorkerNodes().size());
```

Comment:
`nodeManager.getRequiredWorkerNodes().size() * constant`

Reply:
Should we just add taskWriterCount (derived from session in NodePartitioningManager) to getBucketNodeMap SPI and use `nodeManager.getRequiredWorkerNodes().size() * taskWriterCount` here instead of hard coded constant ?

Reply:
I don't think we need to complicate SPI. We could simply assume some constant value here, e.g 32.